src/persist/src/indexed/runtime.rs

```diff
@@ -0,0 +1,233 @@
+                // so log a message. Unfortunately, there isn't really a way to
+                // put the panic message in this log.
+                log::error!("persist runtime thread panic'd");
+            }
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct Runtime {
+    core: Option<RuntimeCore>,
+}
+
+impl Runtime {
+    pub fn start<U: Buffer + Send + 'static, L: Blob + Send + 'static>(
+        indexed: Indexed<U, L>,
+    ) -> Self {
+        let (tx, rx) = mpsc::channel();
+        let handle = thread::spawn(move || {
+            let mut l = RuntimeLoop { indexed, rx };
+            while l.work() {}
+        });
+        let handle = Arc::new(Mutex::new(Some(handle)));
+        let core = RuntimeCore { handle, tx };
+        Runtime { core: Some(core) }
+    }
+
+    pub fn register(&self, id: Id) -> Result<(), Error> {
+        let core = self.get_core("register")?;
+
+        let (tx, rx) = mpsc::channel();
+        let callback = move |res| {
+            tx.send(res).expect("internal error");
+        };
+        let cmd = Cmd::Register(id, Box::new(callback));
+        core.send(cmd);
+
+        rx.recv().expect("WIP")
+    }
+
+    pub fn write_sync(
+        &self,
+        id: Id,
+        updates: &[((String, String), u64, isize)],
+    ) -> Result<(), Error> {
+        let core = self.get_core("write_sync")?;
+
+        let (tx, rx) = mpsc::channel();
+        let callback = move |res| {
+            tx.send(res).expect("internal error");
+        };
+        let cmd = Cmd::Write(id, updates.to_vec(), Box::new(callback));
+        core.send(cmd);
+
+        rx.recv().expect("WIP")
+    }
+
+    pub fn seal(&self, id: Id, ts: u64) -> Result<(), Error> {
+        let core = self.get_core("seal")?;
+
+        let (tx, rx) = mpsc::channel();
+        let callback = move |res| {
+            tx.send(res).expect("internal error");
+        };
+        let cmd = Cmd::Seal(id, ts, Box::new(callback));
+        core.send(cmd);
+
+        rx.recv().expect("WIP")
+    }
+
+    pub fn snapshot(&self, id: Id) -> Result<IndexedSnapshot, Error> {
+        let core = self.get_core("snapshot")?;
+
+        let (tx, rx) = mpsc::channel();
+        let callback = move |res| {
+            tx.send(res).expect("internal error");
+        };
+        let cmd = Cmd::Snapshot(id, Box::new(callback));
+        core.send(cmd);
+
+        rx.recv().expect("WIP")
+    }
+
+    pub fn stop(&mut self) {
+        // This is the only self.core.take call, so if self.core is None, it
+        // means we already stopped.
+        if let Some(core) = self.core.take() {
+            core.stop()
+        }
+    }
+
+    fn get_core(&self, op: &str) -> Result<&RuntimeCore, Error> {
+        self.core
+            .as_ref()
+            .ok_or_else(|| Error::from(format!("operation {} called on a stopped Runtime", op)))
+    }
+}
+
+impl Drop for Runtime {
+    fn drop(&mut self) {
```

Comment:
it seems like there could be multiple clones of `Runtime` - I'm not sure if we want to stop when any of them drop (it seems more natural to stop when the last one drops?)